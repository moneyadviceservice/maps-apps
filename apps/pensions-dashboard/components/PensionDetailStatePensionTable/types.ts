export type Row = {
  title: string;
  monthlyAmount: string;
  yearlyAmount: string;
  payableDate: string;
};
