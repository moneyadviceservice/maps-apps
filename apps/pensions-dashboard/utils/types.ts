export type ErrorResponse = {
  error: string;
  message: string;
};

export type PensionData = {
  pensionPolicies: PensionPolicy[];
  peiInformation: PeiInformationModel;
  pensionsDataRetrievalComplete: boolean;
};

type PensionPolicy = {
  pensionArrangements: PensionArrangement[];
};

export type PensionArrangement = {
  externalPensionPolicyId: string;
  externalAssetId: string;
  matchType: 'DEFN' | 'POSS' | 'CONT' | 'NEW' | 'SYS';
  schemeName: string;
  alternateSchemeNames?: AlternateScheme[];
  contactReference: string;
  pensionType: 'AVC' | 'CB' | 'CDC' | 'DB' | 'DC' | 'HYB' | 'SP';
  pensionOrigin?: 'A' | 'PC' | 'PM' | 'PT' | 'WC' | 'WM' | 'WT';
  pensionStatus?: 'A' | 'I' | 'IPPF' | 'IWU';
  startDate?: string;
  retirementDate?: string;
  dateOfBirth?: string;
  statePensionMessageEng?: string;
  statePensionMessageWelsh?: string;
  contributionsFromMultipleEmployers: boolean;
  pensionAdministrator: PensionAdministrator;
  employmentMembershipPeriods?: EmploymentMembershipPeriod[];
  benefitIllustrations?: BenefitIllustration[];
  additionalDataSources?: AdditionalDataSource[];
};

type AlternateScheme = {
  name: string;
  alternateNameType: 'FOR' | 'OTH';
};

type PensionAdministrator = {
  name: string;
  contactMethods: ContactMethod[];
};

type ContactMethod = {
  preferred: boolean;
  contactMethodDetails: EmailAddress | PhoneNumber | PostalAddress | Website;
};

type EmploymentMembershipPeriod = {
  employerName: string;
  employerStatus: 'C' | 'H';
  membershipStartDate: string;
  membershipEndDate: string;
};

type BenefitIllustration = {
  illustrationComponents: BenefitIllustrationComponent[];
  illustrationDate: string;
};

type AdditionalDataSource = {
  url: string;
  informationType: 'C_AND_C' | 'SIP' | 'IMP' | 'ANR' | 'SP';
};

type PeiInformationModel = {
  peiRetrievalComplete: boolean;
  peiData: PeiData[];
};

type PeiData = {
  pei: string;
  description: string;
  retrievalStatus:
    | 'NEW'
    | 'RETRIEVAL_REQUESTED'
    | 'RETRIEVAL_COMPLETE'
    | 'RETRIEVAL_FAILED';
};

type EmailAddress = {
  email: string;
};

type PhoneNumber = {
  number: string;
  usage: ('A' | 'M' | 'N' | 'S' | 'W')[];
};

type PostalAddress = {
  postalName: string;
  line1: string;
  line2?: string;
  line3?: string;
  line4?: string;
  line5?: string;
  postcode: string;
  countryCode: string;
};

type Website = {
  url: string;
};

type BenefitIllustrationComponent = {
  illustrationType: 'ERI' | 'AP';
  unavailableReason?:
    | 'ANO'
    | 'DB'
    | 'DBC'
    | 'DCC'
    | 'DCHA'
    | 'DCHP'
    | 'DCSM'
    | 'MEM'
    | 'NEW'
    | 'NET'
    | 'PPF'
    | 'TRN'
    | 'WU';
  benefitType: 'AVC' | 'CBL' | 'CBS' | 'CDI' | 'CDL' | 'DB' | 'DBL' | 'DC';
  calculationMethod: 'BS' | 'CBI' | 'SMPI';
  payableDetails: PayableDetails;
  dcPot?: number;
  survivorBenefit?: boolean;
  safeguardedBenefit?: boolean;
  illustratonWarnings?: (
    | 'AVC'
    | 'CUR'
    | 'DEF'
    | 'FAS'
    | 'PEO'
    | 'PNR'
    | 'PSO'
    | 'SPA'
    | 'TVI'
    | 'INP'
  )[];
};

type PayableDetails =
  | RecurringIncomeDetails
  | LumpSumDetails
  | AmountNotProvidedDetails;

type RecurringIncomeDetails = {
  amountType: 'INC' | 'INCL' | 'INCN';
  annualAmount: number;
  monthlyAmount: number;
  lastPaymentDate?: string;
  payableDate: string;
  increasing: boolean;
};

type LumpSumDetails = {
  amountType: 'CSH' | 'CSHL' | 'CSHN';
  amount: number;
  payableDate: string;
};

type AmountNotProvidedDetails = {
  reason: 'SML';
  payableDate: string;
  lastPaymentDate?: string;
};

export type PensionsOverviewModel = {
  totalPensions: number;
  confirmedPensions: PensionArrangement[];
  unconfirmedPensions: PensionArrangement[];
  incompletePensions: PensionArrangement[];
};
