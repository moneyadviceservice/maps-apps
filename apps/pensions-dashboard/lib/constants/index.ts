import Cookies from 'cookies';

export const STATE_RETIREMENT_AGE = 67;

export const NO_DATA = '--';

export const NUMBER_WORDS = [
  { en: 'zero', cy: 'sero' },
  { en: 'one', cy: 'un' },
  { en: 'two', cy: 'dau' },
  { en: 'three', cy: 'tri' },
  { en: 'four', cy: 'pedwar' },
  { en: 'five', cy: 'pump' },
  { en: 'six', cy: 'chwech' },
  { en: 'seven', cy: 'saith' },
  { en: 'eight', cy: 'wyth' },
  { en: 'nine', cy: 'naw' },
  { en: 'ten', cy: 'deg' },
  { en: 'eleven', cy: 'un ar ddeg' },
  { en: 'twelve', cy: 'deuddeg' },
];

export const PROTOCOL =
  process?.env.NODE_ENV === 'development' ? 'http://' : 'https://';

export const COOKIE_OPTIONS = {
  path: '/',
  httpOnly: true,
  sameSite: 'lax',
} as Cookies.SetOption;

export enum PensionsCardType {
  CONFIRMED = 'CONFIRMED',
  CONFIRMED_NO_INCOME = 'CONFIRMED_NO_INCOME',
  PENDING = 'PENDING',
}

export enum MatchType {
  DEFN = 'DEFN',
  POSS = 'POSS',
  CONT = 'CONT',
  NEW = 'NEW',
  SYS = 'SYS',
}

export enum PensionType {
  AVC = 'AVC',
  CB = 'CB',
  CDC = 'CDC',
  DB = 'DB',
  DC = 'DC',
  HYB = 'HYB',
  SP = 'SP',
}

export enum PensionOrigin {
  A = 'A',
  PC = 'PC',
  PM = 'PM',
  PT = 'PT',
  WC = 'WC',
  WM = 'WM',
  WT = 'WT',
}

export enum PensionStatus {
  A = 'A',
  I = 'I',
  IPPF = 'IPPF',
  IWU = 'IWU',
}

export enum AlternateNameType {
  FOR = 'FOR',
  OTH = 'OTH',
}

export enum EmployerStatus {
  C = 'C',
  H = 'H',
}

export enum InformationType {
  C_AND_C = 'C_AND_C',
  SIP = 'SIP',
  IMP = 'IMP',
  ANR = 'ANR',
  SP = 'SP',
}

export enum RetrievalStatus {
  NEW = 'NEW',
  RETRIEVAL_REQUESTED = 'RETRIEVAL_REQUESTED',
  RETRIEVAL_COMPLETE = 'RETRIEVAL_COMPLETE',
  RETRIEVAL_FAILED = 'RETRIEVAL_FAILED',
}

export enum UsageType {
  A = 'A',
  M = 'M',
  N = 'N',
  S = 'S',
  W = 'W',
}

export enum IllustrationType {
  ERI = 'ERI',
  AP = 'AP',
  DB = 'DB',
}

export enum BenefitType {
  AVC = 'AVC',
  CBL = 'CBL',
  CBS = 'CBS',
  CDI = 'CDI',
  CDL = 'CDL',
  DB = 'DB',
  DBL = 'DBL',
  DC = 'DC',
  SP = 'SP',
}

export enum UnavailableReason {
  ANO = 'ANO',
  DB = 'DB',
  DBC = 'DBC',
  DCC = 'DCC',
  DCHA = 'DCHA',
  DCHP = 'DCHP',
  DCSM = 'DCSM',
  MEM = 'MEM',
  NEW = 'NEW',
  NET = 'NET',
  PPF = 'PPF',
  TRN = 'TRN',
  WU = 'WU',
}

export enum IllustrationWarning {
  AVC = 'AVC',
  CUR = 'CUR',
  DEF = 'DEF',
  FAS = 'FAS',
  PEO = 'PEO',
  PNR = 'PNR',
  PSO = 'PSO',
  SPA = 'SPA',
  TVI = 'TVI',
  INP = 'INP',
}

export enum CalculationMethod {
  BS = 'BS',
  CBI = 'CBI',
  SMPI = 'SMPI',
}

export enum LumpSumAmountType {
  CSH = 'CSH',
  CSHL = 'CSHL',
  CSHN = 'CSHN',
}

export enum RecurringAmountType {
  INC = 'INC',
  INCL = 'INCL',
  INCN = 'INCN',
}

export enum AmountNotProvidedReason {
  SML = 'SML',
}

export enum ContactMethods {
  TELEPHONE = 'number',
  EMAIL = 'email',
  POSTAL_NAME = 'postalName',
  WEB_CONTACTS = 'url',
}

export const BACK_LINKS: { [key in string]: string } = {
  CONFIRMED: '/your-pension-breakdown',
  INCOMPLETE: '/pending-pensions',
  UNCONFIRMED: '/pensions-that-need-action',
  '': '/your-pension-search-results',
};

// API response messages
export const REQUEST_ABANDONED = 'Network request was abandoned';
export const RESPONSE_NOT_OK = 'Network response was not ok';
export const RESPONSE_NOT_ACCEPTED = 'Network response was not accepted';
export const REQUEST_FAILED = 'Network request failed';
export const DATA_NOT_FOUND = 'Data not found';

export const DEFAULT_LOAD_TIME = 80;

export enum PensionGroup {
  GREEN = 'green',
  GREEN_NO_INCOME = 'green_no_income',
  RED = 'red',
  YELLOW = 'yellow',
}
