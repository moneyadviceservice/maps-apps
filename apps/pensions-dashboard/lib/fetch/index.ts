export * from './delete-pension-data';
export * from './get-all-pensions';
export * from './get-claims-gathering-redirect';
export * from './get-csrf-token';
export * from './get-load-times';
export * from './get-pension-detail';
export * from './get-pensions-data';
export * from './get-pensions-overview';
export * from './post-pensions-data';
export * from './post-pensions-data-retrieval';
export * from './post-redirect-details';
export * from './validate-access-token';
