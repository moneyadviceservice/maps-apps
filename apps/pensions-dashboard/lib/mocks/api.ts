import {
  MatchType,
  PensionsCategory,
  PensionsRetrievalStatus,
  PensionType,
} from '../constants';
import { PensionArrangement, UserSession } from '../types';
import {
  PensionsByCategory,
  PensionsStatus,
  PensionsSummary,
} from '../types/api-responses';

export const baseUserSession: UserSession = {
  userSessionId: 'test-session-id',
  authorizationCode: 'test-auth-code',
};

export const fullUserSession: UserSession = {
  userSessionId: 'test-session-id',
  authorizationCode: 'test-auth-code',
  sessionStart: '1000000000',
};

export const mockPensionsStatus: PensionsStatus = {
  predictedTotalDataRetrievalTime: 10,
  predictedRemainingDataRetrievalTime: 10,
  pensionsDataRetrievalComplete: true,
};

export const mockPensionsSummary: PensionsSummary = {
  isPensionRetrievalComplete: true,
  pensions: [
    {
      pei: 'pension1',
      schemeName: 'Scheme 1',
      category: PensionsCategory.CONFIRMED,
      pensionType: PensionType.DC,
      hasIncome: false,
      administratorName: 'Admin 1',
      retrievalStatus: PensionsRetrievalStatus.COMPLETED,
    },
    {
      pei: 'pension2',
      schemeName: 'Scheme 2',
      category: PensionsCategory.PENDING,
      pensionType: PensionType.DB,
      hasIncome: false,
      administratorName: 'Admin 2',
      retrievalStatus: PensionsRetrievalStatus.COMPLETED,
    },
    {
      pei: 'pension3',
      schemeName: 'Scheme 3',
      category: PensionsCategory.CONFIRMED,
      pensionType: PensionType.SP,
      hasIncome: true,
      administratorName: 'Admin 3',
      retrievalStatus: PensionsRetrievalStatus.COMPLETED,
    },
    {
      pei: 'pension4',
      schemeName: 'Scheme 4',
      category: PensionsCategory.UNSUPPORTED,
      pensionType: PensionType.AVC,
      hasIncome: false,
      administratorName: 'Admin 4',
      retrievalStatus: PensionsRetrievalStatus.COMPLETED,
    },
    {
      pei: 'pension5',
      schemeName: 'Scheme 5',
      category: PensionsCategory.ERROR,
      pensionType: PensionType.DB,
      hasIncome: false,
      administratorName: 'Admin 5',
      retrievalStatus: PensionsRetrievalStatus.COMPLETED,
    },
  ],
};

export const mockPensionDetailById: PensionArrangement = {
  externalPensionPolicyId: 'pension1',
  externalAssetId: 'asset1',
  matchType: MatchType.DEFN,
  schemeName: 'Scheme 1',
  contactReference: 'Contact-1',
  pensionType: PensionType.DB,
  pensionAdministrator: {
    name: 'Admin 1',
    contactMethods: [
      {
        preferred: true,
        contactMethodDetails: {
          email: 'admin1@example.com',
        },
      },
    ],
  },
  contributionsFromMultipleEmployers: false,
};

export const mockPensionsByCategory: PensionsByCategory = {
  totalContactPensions: 1,
  isPensionRetrievalComplete: true,
  arrangements: [[mockPensionDetailById]],
};
