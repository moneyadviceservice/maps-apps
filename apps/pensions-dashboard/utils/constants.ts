export const SITE_TITLE = 'MoneyHelper Pensions Dashboard';

export const NUMBER_WORDS = [
  { en: 'zero', cy: 'sero' },
  { en: 'one', cy: 'un' },
  { en: 'two', cy: 'dau' },
  { en: 'three', cy: 'tri' },
  { en: 'four', cy: 'pedwar' },
  { en: 'five', cy: 'pump' },
  { en: 'six', cy: 'chwech' },
  { en: 'seven', cy: 'saith' },
  { en: 'eight', cy: 'wyth' },
  { en: 'nine', cy: 'naw' },
  { en: 'ten', cy: 'deg' },
  { en: 'eleven', cy: 'un ar ddeg' },
  { en: 'twelve', cy: 'deuddeg' },
];

export const PROTOCOL =
  process?.env.NODE_ENV === 'development' ? 'http://' : 'https://';
