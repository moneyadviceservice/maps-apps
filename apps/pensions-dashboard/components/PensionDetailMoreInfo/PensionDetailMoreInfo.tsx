import { Link } from '@maps-react/common/components/Link';
import { Paragraph } from '@maps-react/common/components/Paragraph';
import useTranslation from '@maps-react/hooks/useTranslation';

import { InformationType } from '../../lib/constants';
import { AdditionalDataSource, PensionArrangement } from '../../lib/types';

type PensionDetailMoreInfoProps = {
  data: PensionArrangement;
};

export const PensionDetailMoreInfo = ({ data }: PensionDetailMoreInfoProps) => {
  const { t } = useTranslation();

  const getMoreInfoMessage = ({ informationType }: AdditionalDataSource) => {
    switch (informationType) {
      case InformationType.C_AND_C:
        return t('pages.pension-details.more-info.costs');
      case InformationType.SP:
        return t('pages.pension-details.more-info.sp');
      default:
        return '';
    }
  };

  return data.additionalDataSources?.map((additionalData, idx) => {
    return (
      <Paragraph key={`${additionalData.informationType}-${idx}`}>
        {getMoreInfoMessage(additionalData)}{' '}
        <Link asInlineText target="_blank" href={additionalData.url}>
          {additionalData.url}
        </Link>
        .
      </Paragraph>
    );
  });
};
