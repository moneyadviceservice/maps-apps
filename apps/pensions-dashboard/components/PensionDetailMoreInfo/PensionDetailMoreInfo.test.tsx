import { render } from '@testing-library/react';

import useTranslation from '@maps-react/hooks/useTranslation';

import { InformationType } from '../../lib/constants';
import { mockPensionsData } from '../../lib/mocks';
import { PensionArrangement } from '../../lib/types';
import { PensionDetailMoreInfo } from './PensionDetailMoreInfo';

import '@testing-library/jest-dom/extend-expect';

jest.mock('@maps-react/hooks/useTranslation');

const mockUseTranslation = useTranslation as jest.Mock;

const mockData = mockPensionsData.pensionPolicies[0]
  .pensionArrangements[0] as PensionArrangement;

describe('PensionDetailMoreInfo', () => {
  beforeEach(() => {
    mockUseTranslation.mockReturnValue({
      t: (key: string) => key,
    });
  });

  it('renders correctly', () => {
    const { container } = render(<PensionDetailMoreInfo data={mockData} />);
    expect(container).toMatchSnapshot();
  });

  it('renders the correct information for C_AND_C type', () => {
    const data = {
      ...mockData,
      additionalDataSources: [
        {
          informationType: InformationType.C_AND_C,
          url: 'http://example.com/costs',
        },
      ],
    };
    const { getByText } = render(<PensionDetailMoreInfo data={data} />);
    expect(
      getByText(/pages.pension-details.more-info.costs/),
    ).toBeInTheDocument();
    expect(getByText('(opens in a new window)')).toBeInTheDocument();
    expect(getByText('http://example.com/costs')).toBeInTheDocument();
  });

  it('renders the correct information for SP type', () => {
    const data = {
      ...mockData,
      additionalDataSources: [
        {
          informationType: InformationType.SP,
          url: 'http://example.com/sp',
        },
      ],
    };
    const { getByText } = render(<PensionDetailMoreInfo data={data} />);
    expect(getByText(/pages.pension-details.more-info.sp/)).toBeInTheDocument();
    expect(getByText('http://example.com/sp')).toBeInTheDocument();
  });

  it('renders nothing if additionalDataSources is empty', () => {
    const data: PensionArrangement = {
      ...mockData,
      additionalDataSources: [],
    };
    const { container } = render(<PensionDetailMoreInfo data={data} />);
    expect(container).toBeEmptyDOMElement();
  });
});
