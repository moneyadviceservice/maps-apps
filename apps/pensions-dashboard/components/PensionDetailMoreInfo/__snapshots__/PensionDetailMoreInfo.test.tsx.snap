// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PensionDetailMoreInfo renders correctly 1`] = `<div />`;
